docs/zh/config-desc.md
# 配置文件

默认会在 ~/.sophon-auth/config.tml 生成配置文件

```toml
# 服务使用端口,提供HTTP服务
Listen = "127.0.0.1:8989"
ReadTimeout = "1m"
WriteTimeout = "1m"
# IdleTimeout is the maximum amount of time to wait for the
# next request when keep-alives are enabled. If IdleTimeout
# is zero, the value of ReadTimeout is used. If both are
# zero, there is no timeout.
IdleTimeout = "1m"

[db]
  # 支持: badger (默认), mysql
  type = "badger"
  # 以下参数适用于MySQL
  DSN = "rennbon:111111@(127.0.0.1:3306)/auth_server?parseTime=true&loc=Local&charset=utf8mb4&collation=utf8mb4_unicode_ci&readTimeout=10s&writeTimeout=10s"
  # conns 1500 concurrent
  maxOpenConns = 64
  maxIdleConns = 128
  maxLifeTime = "120s"
  maxIdleTime = "30s"
  debug = false

# 日志默认写入std
[log]
  # trace,debug,info,warning,error,fatal,panic
  # 默认日志级别
  logLevel = trace
  # db type, 默认：0， 1:influxDB，暂时没用，无需理会
  type = 0
  # db hook switch，是否使用influxdb
  hookSwitch = false 

# 可选; 日志数据库
[log.influxdb]
serverURL = "http://192.168.1.141:8086"
authToken = "jcomkQ-dVBRoCrKSEWMuYxA4COj_EfyCvwgPW5Ql-tT-cCizIjE24rPJQNx8Kkqzz4gCW8YNFq0wcDaHJOcGMQ=="
org = "venus-oauth"
bucket = "bkt2"
measurement = "verify"
flushInterval = "30s"
batchSize = 100

# 可选
[Trace]
  # 是否启用 trace
  JaegerTracingEnabled = true
  # 收集的频率
  ProbabilitySampler = 1.0
  JaegerEndpoint = "127.0.0.1:6831"
  ServerName = "sophon-auth"
```

CHANGELOG.md
# sophon-auth changelog

## v1.12.0

### New Features
* feat: add flag --listen and default listening 127.0.0.1  https://github.com/ipfs-force-community/sophon-auth/pull/167

### Optimize
* opt: remove perm adapt strategy  https://github.com/ipfs-force-community/sophon-auth/pull/164
* opt: Remove the configuration field secret  https://github.com/ipfs-force-community/sophon-auth/pull/165
* opt: add default config for mysql  https://github.com/ipfs-force-community/sophon-auth/pull/168
* opt: Adjust the order of the permissions array  https://github.com/ipfs-force-community/sophon-auth/pull/169
* opt: rebranding from venus-auth to sophon-auth  https://github.com/ipfs-force-community/sophon-auth/pull/170
* opt: rm perm chack falg / 移除兼容权限检查的 flag  https://github.com/ipfs-force-community/sophon-auth/pull/171

## v1.11.0
### New Features
* feat: add status api to detect api ready / 添加状态检测接口 [[#144](https://github.com/ipfs-force-community/sophon-auth/pull/144)]
* feat: use thirty party healthcheck lib  / 添加healthcheck接口 [[#145](https://github.com/ipfs-force-community/sophon-auth/pull/145)]
* feat: add api protection / 增加接口保护  [[#140](https://github.com/ipfs-force-community/sophon-auth/pull/140)]
* feat: return detailed error infor after authentication failure  /鉴权失败后返回更加详细的错误信息 [[#160](https://github.com/ipfs-force-community/sophon-auth/pull/160)]
* feat: add docker push / 增加推送到镜像仓库的功能 [[#161](https://github.com/ipfs-force-community/sophon-auth/pull/161)]

### Bug Fixes

* fix: repo not exist by 修复启动时目录不存在从而启动失败的问题 [[#157](https://github.com/ipfs-force-community/sophon-auth/pull/157)]
* fix: not set flag value to config  /修复配置错误 [[#158](https://github.com/ipfs-force-community/sophon-auth/pull/158)]
* fix: cli not found config  / 修复创建目录失败的问题 [[#159](https://github.com/ipfs-force-community/sophon-auth/pull/159)]


## v1.10.1

* 查询参数为空时不重置请求url [[#153](https://github.com/ipfs-force-community/sophon-auth/pull/153)]
* 补充对 delegated 地址的支持 [[#154](https://github.com/ipfs-force-community/sophon-auth/pull/154)]

## v1.10.0

* 升级 go-jsonrpc 到 v0.1.7

## v1.10.0-rc2

* 简化 authClient 接口，并增加 context [[#126](https://github.com/ipfs-force-community/sophon-auth/pull/126)]
* 重写 url 中的地址参数 [[#127](https://github.com/ipfs-force-community/sophon-auth/pull/127)]
* 增加用户数据隔离的工具 [[#130](https://github.com/ipfs-force-community/sophon-auth/pull/130)]
* 调整 jwtclient.IAtuhClient 接口 [[#137](https://github.com/ipfs-force-community/sophon-auth/pull/137)]

## v1.10.0-rc1

* github action 增加 dispatch 事件 [[#138](https://github.com/ipfs-force-community/sophon-auth/pull/138)]
* 升级 github.com/gin-gonic/gin 版本，从 1.7.0 升级到 1.7.7 [[#139](https://github.com/ipfs-force-community/sophon-auth/pull/139)]
* 升级 github.com/prometheus/client_golang 版本，从 1.11.0 升级到 1.11.1 [[#141](https://github.com/ipfs-force-community/sophon-auth/pull/141)]


config/config.go
package config

import (
	"bytes"
	"crypto/rand"
	"io"
	"os"
	"time"

	"github.com/BurntSushi/toml"
	"github.com/ipfs-force-community/metrics"
	"golang.org/x/xerrors"
)

type Config struct {
	Listen       string               `json:"listen"`
	ReadTimeout  time.Duration        `json:"readTimeout"`
	WriteTimeout time.Duration        `json:"writeTimeout"`
	IdleTimeout  time.Duration        `json:"idleTimeout"`
	Log          *LogConfig           `json:"log"`
	DB           *DBConfig            `json:"db"`
	Trace        *metrics.TraceConfig `json:"traceConfig"`
}

type DBType = string

const (
	Mysql  DBType = "mysql"
	Badger DBType = "badger"
)

type DBConfig struct {
	Type         DBType        `json:"type"`
	DSN          string        `json:"dsn"`
	MaxOpenConns int           `json:"maxOpenConns"`
	MaxIdleConns int           `json:"maxIdleConns"`
	MaxLifeTime  time.Duration `json:"maxLifeTime"`
	MaxIdleTime  time.Duration `json:"maxIdleTime"`
	Debug        bool          `json:"debug"`
}

// RandSecret If the daemon does not have a secret key configured, it is automatically generated
func RandSecret() ([]byte, error) {
	sk, err := io.ReadAll(io.LimitReader(rand.Reader, 32))
	if err != nil {
		return nil, err
	}
	return sk, nil
}

func DefaultConfig() *Config {
	return &Config{
		Listen:       "127.0.0.1:8989",
		ReadTimeout:  time.Minute,
		WriteTimeout: time.Minute,
		IdleTimeout:  time.Minute,
		Trace: &metrics.TraceConfig{
			JaegerTracingEnabled: false,
			ProbabilitySampler:   1.0,
			JaegerEndpoint:       "localhost:6831",
			ServerName:           "sophon-auth",
		},
		Log: &LogConfig{
			LogLevel:   "trace",
			HookSwitch: false,
		},
		DB: &DBConfig{
			Type:         Badger,
			MaxOpenConns: 64,
			MaxIdleConns: 128,
			MaxLifeTime:  120 * time.Second,
			MaxIdleTime:  60 * time.Second,
		},
	}
}

type LogHookType = int

const (
	LHTInfluxDB LogHookType = 1
)

type LogConfig struct {
	LogLevel   string          `json:"logLevel"`
	Type       LogHookType     `json:"type"`
	HookSwitch bool            `json:"hookSwitch"`
	InfluxDB   *InfluxDBConfig `json:"influxdb"`
}

type InfluxDBConfig struct {
	ServerURL     string        `json:"serverURL"`
	AuthToken     string        `json:"authToken"`
	Org           string        `json:"org"`
	Bucket        string        `json:"bucket"`
	FlushInterval time.Duration `json:"flushInterval"`
	BatchSize     uint          `json:"batchSize"`
}

func MakeDir(path string) error {
	err := os.Mkdir(path, 0o755)
	if err != nil && !os.IsExist(err) {
		return err
	}
	return nil
}

func DecodeConfig(path string) (c *Config, err error) {
	provider, err := FromConfigString(path, "toml")
	if err != nil {
		return nil, err
	}
	c = new(Config)
	err = provider.Unmarshal(c)
	if err != nil {
		return nil, err
	}
	return
}

func Cover(path string, config *Config) error {
	c, err := os.Create(path)
	if err != nil {
		return err
	}
	barr, err := config.toBytes()
	if err != nil {
		return err
	}
	_, err = c.Write(barr)
	if err != nil {
		return xerrors.Errorf("write config: %w", err)
	}
	if err := c.Close(); err != nil {
		return xerrors.Errorf("close config: %w", err)
	}
	return nil
}

func (c *Config) toBytes() ([]byte, error) {
	buf := new(bytes.Buffer)
	e := toml.NewEncoder(buf)
	if err := e.Encode(c); err != nil {
		return nil, xerrors.Errorf("encoding config: %w", err)
	}
	b := buf.Bytes()
	b = bytes.ReplaceAll(b, []byte("#["), []byte("["))
	return b, nil
}
